### PR TITLE
setup stats command

### PR DESCRIPTION
**Pull request by fahedouch**
Signed-off-by: fahed dorgaa <fahed.dorgaa@gmail.com>

### CODE REVIEW

File: stats.go [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	//"encoding/json"
```
Comment (AkihiroSuda) [inline stats.go:20]: Could you clean up comment lines

File: stats.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
+/*
+   Copyright (C) nerdctl authors.
```
Comment (AkihiroSuda) [inline stats.go:2]: This file seems derivative of containerd, right?

File: formatter_stats.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+const (
+	winOSType                  = "windows"
+	defaultStatsTableFormat    = "table {{.ID}}\t{{.Name}}\t{{.CPUPerc}}\t{{.MemUsage}}\t{{.MemPerc}}\t{{.NetIO}}\t{{.BlockIO}}\t{{.PIDs}}"
+	winDefaultStatsTableFormat = "table {{.ID}}\t{{.Name}}\t{{.CPUPerc}}\t{{.MemUsage}}\t{{.NetIO}}\t{{.BlockIO}}"
+
+	containerHeader = "CONTAINER"
+	cpuPercHeader   = "CPU %"
+	netIOHeader     = "NET I/O"
+	blockIOHeader   = "BLOCK I/O"
+	memPercHeader   = "MEM %"             // Used only on Linux
```
Comment (AkihiroSuda) [inline formatter_stats.go:21]: I'd prefer not to add Windows-specific code at the moment (until we support `nerdctl run` on Windows)

File: formatter_stats.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
+package main
```
Comment (AkihiroSuda) [inline formatter_stats.go:1]: Can we have this under a separate pkg like `pkg/stats`?

File: stats.go (hunk 2 above)
Comment (fahedouch) [reply stats.go:2]: not 100%, several logic I introduced them myself

Comment (fahedouch) [reply stats.go:2]: You want to add containerd Copyright ?

File: go.mod [hunk 5]
```diff
@@ -18,8 +19,10 @@ require (
 	github.com/fvbommel/sortorder v1.0.2 // indirect
 	github.com/gogo/protobuf v1.3.1
 	github.com/google/go-cmp v0.5.4 // indirect
+	github.com/jaguilar/vt100 v0.0.0-20201024211400-81de19cb81a4 // indirect
 	github.com/lib/pq v1.9.0 // indirect
 	github.com/mattn/go-isatty v0.0.12
+	github.com/moby/buildkit v0.8.1 // indirect
```
Comment (AkihiroSuda) [inline go.mod:25]: Why?

File: go.mod [hunk 6]
```diff
@@ -32,15 +35,27 @@ require (
 	github.com/spf13/viper v1.7.1 // indirect
 	github.com/stretchr/testify v1.7.0 // indirect
 	github.com/theupdateframework/notary v0.7.0 // indirect
+	github.com/tonistiigi/go-rosetta v0.0.0-20201102221648-9ba854641817 // indirect
+	github.com/urfave/cli v1.22.2
```
Comment (AkihiroSuda) [inline go.mod:39]: Why?

File: pkg/stats/formatter_stats.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+	cs.NetworkTx = 0
+	cs.BlockRead = 0
+	cs.BlockWrite = 0
+	cs.PidsCurrent = 0
+	cs.err = err
+	cs.IsInvalid = true
+}
+
+// SetError sets container statistics error
+func (cs *Stats) SetError(err error) {
```
Comment (AkihiroSuda) [inline pkg/stats/formatter_stats.go:75]: Unused function?

File: pkg/stats/formatter_stats.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+// This is used to determine whether the statistics are valid or not
+func (cs *Stats) GetError() error {
+	cs.mutex.Lock()
+	defer cs.mutex.Unlock()
+	return cs.err
+}
+
+// SetErrorAndReset zeroes all the container statistics and store the error.
+// It is used when receiving time out error during statistics collecting to reduce lock overhead
+func (cs *Stats) SetErrorAndReset(err error) {
```
Comment (AkihiroSuda) [inline pkg/stats/formatter_stats.go:58]: Unused function?

File: stats.go (hunk 2 above)
Comment (AkihiroSuda) [reply stats.go:2]: Yes

File: stats.go [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+		},
+	},
+	Action: func(clicontext *cli.Context) error {
+		optionsStats.containers = clicontext.Args().Slice()
+		showAll := len(optionsStats.containers) == 0
+		closeChan := make(chan error)
+		var err error
+
+		if runtime.GOOS == "windows" {
+			fmt.Println("nerdctl do not yet support windows")
```
Comment (AkihiroSuda) [inline stats.go:83]: This is not specific to `stats`, and I don't even think it compiles on Windows, so this Println is not needed

File: stats.go [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+				u <- err
+			}
+
+			switch v := anydata.(type) {
+			case *v1.Metrics:
+				data = v
+			case *v2.Metrics:
+				data2 = v
+			default:
+				//return errors.New("cannot convert metric data to cgroups.Metrics or windows.Statistics")
```
Comment (AkihiroSuda) [inline stats.go:259]: Please clean up comments

File: stats.go [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+			if data != nil {
+				cpuPercent = calculateCPUPercentUnixV1(previousCPU, previousSystem, data)
+				blkRead, blkWrite = calculateBlockIO(data)
+				mem = calculateMemUsageUnixNoCache(data)
+				memLimit = float64(data.Memory.Usage.Limit)
+				memPercent = calculateMemPercentUnixNoCache(memLimit, mem)
+				pidsStatsCurrent = data.Pids.Current
+				netRx, netTx = calculateNetwork(data)
+			} else if data2 != nil {
+				//previousCPU = data2.CPU.UsageUsec
```
Comment (AkihiroSuda) [inline stats.go:295]: Is this WIP?

File: stats.go [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+			close(closeChan)
+
+			// make sure each container get at least one valid stat data
+			waitFirst.Wait()
+
+		}
+
+		format := optionsStats.format
+		if len(format) == 0 {
+			/*if len(dockerCli.ConfigFile().StatsFormat) > 0 {
```
Comment (AkihiroSuda) [inline stats.go:156]: Please clean up comments

File: pkg/stats/formatter_stats.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%s / %s", units.HumanSizeWithPrecision(c.s.BlockRead, 3), units.HumanSizeWithPrecision(c.s.BlockWrite, 3))
+}
+
+func (c *statsContext) PIDs() string {
+	if c.s.IsInvalid {
+		return fmt.Sprintf("--")
+	}
+	return fmt.Sprintf("%d", c.s.PidsCurrent)
```
Comment (AkihiroSuda) [inline pkg/stats/formatter_stats.go:213]: Should take clicontext.App.Writer

**Comment by AkihiroSuda** [comment]
Thanks for working on this, but needs cleaning up the code, squashing the commits, and rebasing with the master.

**Comment by AkihiroSuda** [comment]
@fahedouch Are you still planning to work on this?

**Comment by fahedouch** [comment]
@AkihiroSuda yes, if you see if it's a priority I can prioritize it over other tasks

**Comment by nehaldattani** [comment]
This enhancement is pending for quite a while.   Would this PR be accepted anytime soon ?

**Comment by fahedouch** [comment]
I did the necessary except the stats for the `windows container`; some statistics are missing, I am trying to find out how to do the formulas to have the right stats for windows containers

File: cmd/nerdctl/stats.go [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"text/tabwriter"
+	"text/template"
+	"time"
+
+	wstats "github.com/Microsoft/hcsshim/cmd/containerd-shim-runhcs-v1/stats"
```
Comment (AkihiroSuda) [inline cmd/nerdctl/stats.go:30]: Can we split this to `*_windows.go` and `*_linux.go`?

**Comment by AkihiroSuda** [comment]
Thanks for working on this, but this doesn't seem to show containers that are created after starting `nerdctl stats`.

Windows metrics can be added later. (We don't even have `nerdctl run` for Windows yet)

**Comment by AkihiroSuda** [comment]
Also, I see strange CPU stat like `11496300.00%`.
(Only tested cgroup v2, rootless)

File: cmd/nerdctl/stats.go [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+	var statsCommand = &cobra.Command{
+		Use:               "stats",
+		Short:             "Display a live stream of container(s) resource usage statistics.",
+		RunE:              statsAction,
+		ValidArgsFunction: statsShellComplete,
+		SilenceUsage:      true,
+		SilenceErrors:     true,
+	}
+
+	addStastFlags(statsCommand)
```
Comment (dcantah) [inline cmd/nerdctl/stats.go:56]: stats*

File: cmd/nerdctl/stats.go [hunk 16]
```diff
@@ -0,0 +1,427 @@
...
+
+	return statsCommand
+}
+
+func addStastFlags(cmd *cobra.Command) {
```
Comment (dcantah) [inline cmd/nerdctl/stats.go:61]: stats*

File: cmd/nerdctl/stats.go [hunk 17]
```diff
@@ -0,0 +1,427 @@
...
+	}
+	return -1, false
+}
+
+func statsAction(cmd *cobra.Command, args []string) error {
+
+	// NOTE: rootless container does not rely on cgroupv1.
+	// more details about possible ways to resolve this concern: #223
+	if rootlessutil.IsRootless() && infoutil.CgroupsVersion() == "1" {
+		return fmt.Errorf("stats requires cgroup v2 for rootless containers, see https://rootlesscontaine.rs/getting-started/common/cgroup2/")
```
Comment (dcantah) [inline cmd/nerdctl/stats.go:99]: Nit: errors.New if no formatting needed

File: pkg/statsutil/stats.go [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+	MemUsage string
+	MemPerc  string
+	NetIO    string
+	BlockIO  string
+	PIDs     string
+}
+
+// Stats represents an entity to store containers statistics synchronously
+type Stats struct {
+	mutex sync.Mutex
```
Comment (dcantah) [inline pkg/statsutil/stats.go:62]: Would an RWMutex be preferred here? Seems it would benefit the Get methods

File: pkg/statsutil/stats.go [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+	return float64(metrics.Memory.Usage)
+}
+
+func calculateMemPercent(limit float64, usedNo float64) float64 {
+	// Limit will never be 0 unless the container is not running and we haven't
+	// got any data from cgroup
+	if limit != 0 {
+		return usedNo / limit * 100.0
+	}
+	return 5
```
Comment (dcantah) [inline pkg/statsutil/stats.go:227]: Why 5?

File: cmd/nerdctl/stats.go [hunk 20]
```diff
@@ -0,0 +1,427 @@
...
+				data2        *v2.Metrics
+				windowsStats *wstats.WindowsContainerStatistics
+			)
+			switch v := anydata.(type) {
+			case *v1.Metrics:
+				data = v
+			case *v2.Metrics:
+				data2 = v
+			case *wstats.WindowsContainerStatistics:
+				windowsStats = v
```
Comment (dcantah) [inline cmd/nerdctl/stats.go:358]: Part of me thinks it's odd to add Windows logic at the moment, as I don't believe this supports running containers Windows yet.

File: pkg/statsutil/stats.go (hunk 19 above)
Comment (fahedouch) [reply pkg/statsutil/stats.go]: fixed

File: pkg/statsutil/stats.go (hunk 18 above)
Comment (fahedouch) [reply pkg/statsutil/stats.go]: thanks, fixed

File: cmd/nerdctl/stats.go (hunk 17 above)
Comment (fahedouch) [reply cmd/nerdctl/stats.go]: fixed

File: cmd/nerdctl/stats.go (hunk 16 above)
Comment (fahedouch) [reply cmd/nerdctl/stats.go]: fixed

File: cmd/nerdctl/stats.go (hunk 15 above)
Comment (fahedouch) [reply cmd/nerdctl/stats.go]: fixed

**Comment by fahedouch** [comment]
> Thanks for working on this, but this doesn't seem to show containers that are created after starting `nerdctl stats`.
> 
> Windows metrics can be added later. (We don't even have `nerdctl run` for Windows yet)

 @AkihiroSuda  as I mentionned in this comment https://github.com/containerd/nerdctl/pull/73/files#diff-bab6d6355f721a47c2a974e88b96dde11fdeaa6a8d862ee3f7974558b00abc52R150 not yet supported. I will add it second PR if it is possible! 

By the way I fixed the Cpu stats for Cgroupv2

File: cmd/nerdctl/stats_linux.go [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+			s := statsutil.NewStats(c.ID())
+			if cStats.add(s) {
+				waitFirst.Add(1)
+				go collect(cmd, s, waitFirst, c.ID(), !noStream)
+			}
+		}
+	}
+
+	if showAll {
+		// TODO: watches for container creation and removal (dynamic stats)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/stats_linux.go:150]: Could you add this limitation to the `--help` text and `README.md`

Comment (AkihiroSuda) [reply cmd/nerdctl/stats_linux.go:150]: And open an issue ticket

**Comment by AkihiroSuda** [comment]
CI failing for Windows and FreeBSD

**Comment by fahedouch** [comment]
@AkihiroSuda  I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.

File: pkg/eventutil/eventutil.go [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+	w.mu.Unlock()
+}
+
+// Watch ranges over the passed in event chan and processes the events based on the
+// handlers created for a given action.
+// To stop watching, close the event chan.
+func (w *eventHandler) Watch(c <-chan *events.Envelope) {
+	for e := range c {
+		w.mu.Lock()
+		fmt.Println(e.Topic)
```
Comment (AkihiroSuda) [inline pkg/eventutil/eventutil.go:48]: remove

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.

Thanks, but creation doesn't seem handled yet

File: pkg/statsutil/stats_linux.go [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func SetCgroup2StatsFields(previousCgroup2CPU, previousCgroup2System uint64, metrics *v2.Metrics) (StatsEntry, error) {
+
+	cpuPercent = calculateCgroup2CPUPercent(previousCgroup2CPU, previousCgroup2System, metrics)
+	blkRead, blkWrite = calculateCgroup2IO(metrics)
+	mem = calculateCgroup2MemUsage(metrics)
+	memLimit = float64(metrics.Memory.UsageLimit)
+	memPercent = calculateMemPercent(memLimit, mem)
+	pidsStatsCurrent = metrics.Pids.Current
```
Comment (AkihiroSuda) [inline pkg/statsutil/stats_linux.go:65]: NetworkTx/Rx missing?

**Comment by fahedouch** [comment]
> > @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.
> 
> Thanks, but creation doesn't seem handled yet

creation is handled only with `--all` flag,  have you tested with with `--all` flag ?

**Comment by AkihiroSuda** [comment]
> > > @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.
> > 
> > 
> > Thanks, but creation doesn't seem handled yet
> 
> @AkihiroSuda creation is handled only with `--all` flag, have you tested with with `--all` flag ?

Can we watch creation events by default?
`docker stats` doesn't need `--all` to watch creation events.

File: pkg/statsutil/stats_linux.go (hunk 23 above)
Comment (fahedouch) [reply pkg/statsutil/stats_linux.go:65]: @AkihiroSuda  it seems that network is not supported by cgroupv2 https://pkg.go.dev/github.com/containerd/cgroups@v1.0.2/v2/stats#Metrics

**Comment by fahedouch** [comment]
> > > > @AkihiroSuda I added the support for dynamic container creation/deletion. I will fix the CI error this night or tomorrow. Sorry for the delay I am traveling abroad.
> > > 
> > > 
> > > Thanks, but creation doesn't seem handled yet
> > 
> > 
> > @AkihiroSuda creation is handled only with `--all` flag, have you tested with with `--all` flag ?
> 
> Can we watch creation events by default? `docker stats` doesn't need `--all` to watch creation events.

fixed @AkihiroSuda

File: pkg/statsutil/stats_linux.go (hunk 23 above)
Comment (AkihiroSuda) [reply pkg/statsutil/stats_linux.go:65]: But `docker stats` work with cgroup v2.
We can revisit this in another PR.

File: README.md [hunk 24]
```diff
@@ -931,6 +931,12 @@ Flags:
 - :whale: `-f, --format`: Format the output using the given Go template, e.g, `{{json .}}`
 
 ## Stats
+### :whale: nerdctl stats
+Display a live stream of container(s) resource usage statistics.
+
+
+Usage: `nerdctl stats [flags]`
```
Comment (AkihiroSuda) [inline README.md:938]: Next time please add `flags` too

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Review by AkihiroSuda — APPROVED** [review]

File: pkg/statsutil/stats_linux.go (hunk 23 above)
Comment (fahedouch) [reply pkg/statsutil/stats_linux.go:65]: @AkihiroSuda  I'm assuming that `docker` doesn't consume the cgroup api from containerd  https://github.com/containerd/cgroups. 
do you think it is a lack in the containerd cgroup api and that I have to open a ticket at  https://github.com/containerd/cgroups

Comment (AkihiroSuda) [reply pkg/statsutil/stats_linux.go:65]: docker scans the stats from netlink, not from cgroups

Comment (fahedouch) [reply pkg/statsutil/stats_linux.go:65]: @AkihiroSuda thank you for checking this I will fix it.

**Pull request merged**